src/Microsoft.Framework.Caching.SqlServer/MonoDatabaseOperations.cs

```diff
@@ -0,0 +1,291 @@
+            {
+                var command = new SqlCommand(SqlQueries.GetCacheItem, connection);
+                command.Parameters
+                    .AddCacheItemId(key)
+                    .AddWithValue("UtcNow", SqlDbType.DateTime, utcNow.UtcDateTime);
+
+                await connection.OpenAsync();
+
+                var reader = await command.ExecuteReaderAsync(
+                    CommandBehavior.SingleRow | CommandBehavior.SingleResult);
+
+                if (await reader.ReadAsync())
+                {
+                    var id = reader.GetString(Columns.Indexes.CacheItemIdIndex);
+
+                    expirationTime = DateTimeOffset.Parse(reader[Columns.Indexes.ExpiresAtTimeIndex].ToString());
+
+                    if (!await reader.IsDBNullAsync(Columns.Indexes.SlidingExpirationInTicksIndex))
+                    {
+                        slidingExpiration = TimeSpan.FromTicks(
+                            Convert.ToInt64(reader[Columns.Indexes.SlidingExpirationInTicksIndex].ToString()));
+                    }
+
+                    if (!await reader.IsDBNullAsync(Columns.Indexes.AbsoluteExpirationIndex))
+                    {
+                        absoluteExpiration = DateTimeOffset.Parse(
+                            reader[Columns.Indexes.AbsoluteExpirationIndex].ToString());
+                    }
+
+                    if (includeValue)
+                    {
+                        value = (byte[])reader[Columns.Indexes.CacheItemValueIndex];
+                    }
+                }
+                else
+                {
+                    return null;
+                }
+            }
+
+            await UpdateCacheItemExpirationAsync(
+                key, utcNow, expirationTime, slidingExpiration, absoluteExpiration);
+
+            return value;
+        }
+
+        public override void SetCacheItem(string key, byte[] value, DistributedCacheEntryOptions options)
+        {
+            var utcNow = SystemClock.UtcNow;
+
+            var expirationInfo = CacheItemExpiration.GetExpirationInfo(utcNow, options);
+
+            using (var connection = new SqlConnection(ConnectionString))
+            {
+                var upsertCommand = new SqlCommand(SqlQueries.SetCacheItem, connection);
+                upsertCommand.Parameters
+                    .AddCacheItemId(key)
+                    .AddCacheItemValue(value)
+                    .AddExpiresAtTimeMono(expirationInfo.ExpiresAtTime)
+                    .AddSlidingExpirationInTicks(expirationInfo.SlidingExpiration)
+                    .AddAbsoluteExpirationMono(expirationInfo.AbsoluteExpiration);
+
+                connection.Open();
+
+                try
+                {
+                    upsertCommand.ExecuteNonQuery();
+                }
+                catch (SqlException ex)
+                {
+                    if (ex.Message.IndexOf(DuplicateKeyErrorText, StringComparison.OrdinalIgnoreCase) > 0)
```

Comment:
Sure, I updated the PR now where we can check a message number which seem to be the one popularly suggested.

I suppose in the following table the message-id and language-id both combined are unique and so is safe to use this message code for different languages, but i could be wrong as this is not a table but a sql view. @divega 

![image](https://cloud.githubusercontent.com/assets/6559784/8443029/ea4fb398-1f37-11e5-870c-17ded0c4ef8c.png)